src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1242,19 +1246,191 @@
+                                                             :env (:env beta-node)
+                                                             :msg "compiling accumulator"}})
+
+                                 (:join-filter-expressions beta-node)
+                                 (handle-expr (compile-join-filter (:join-filter-expressions beta-node)
+                                                                   (:join-filter-join-bindings beta-node)
+                                                                   (:new-bindings beta-node)
+                                                                   (:env beta-node))
+                                              id
+                                              :join-filter-expr
+                                              {:compile-ctx {:condition condition
+                                                             :join-filter-expressions (:join-filter-expressions beta-node)
+                                                             :env (:env beta-node)
+                                                             :msg "compiling accumulate with join filter node"}}))
+            :query prev)))
+      exprs
+      (:id-to-condition-node beta-graph))))
+
+(sc/defn compile-exprs :- schema/NodeFnLookup
+  "Takes a map in form produced by extract-exprs and evaluates the values(expressions) of the map in a batched manner.
+   This allows the eval calls to be more effecient, rather than evaluating each expression on its own."
+  [key->expr :- schema/NodeExprLookup
+   partition-size :- sc/Int]
+  (let [batching-try-eval (fn [node-keys exprs]
+                            ;; Try to evaluate all of the expressions as a batch. If they fail the batch eval then we
+                            ;; try them one by one with their compilation context, this is slow but we were going to fail
+                            ;; anyway.
+                            (try
+                              (eval exprs)
+                              (catch Exception e
+                                (mapv (fn [expr node-keys]
+                                        (let [cmeta (meta node-keys)]
+                                          (with-bindings
+                                            {#'*compile-ctx* (:compile-ctx cmeta)
+                                             #'*file* (:file cmeta *file*)}
+                                            (try-eval expr))))
+                                      exprs
+                                      node-keys)
+                                ;; If none of the rules are the issue, it is likely that the
+                                ;; size of the code trying to be evaluated has exceeded the limit
+                                ;; set by java.
+                                (throw (ex-info "The batching size of expressions has exceeded the limit set by java."
```

Comment:
`ex-info` has an 3-arity version designed to take the `cause`, ie `ex-info [msg map cause]`

So 
```clj
(ex-info "The batching size of expressions has exceeded the limit set by java." {} e)

```

I'm not sure the `:original-exception` will add value. Perhaps if that exception has `ex-data` to include. Also, it may be useful to put the `node-keys` in this map.